bdk_cbf_example/src/main.rs

```diff
@@ -0,0 +1,158 @@
+use bdk_cli::{
+    anyhow,
+    clap::{self, Args, Subcommand, ValueEnum},
+};
+
+use crate::cbf::CbfEvent;
+use bdk_chain::chain_graph::ChainGraph;
+use bdk_chain::keychain::{DerivationAdditions, KeychainChangeSet};
+use bdk_chain::{BlockId, TxHeight};
+
+mod cbf;
+
+#[derive(Args, Debug, Clone)]
+struct CbfArgs {
+    #[arg(value_enum, default_value_t = Domains::IPv4)]
+    domains: Domains,
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, ValueEnum)]
+pub enum Domains {
+    IPv4,
+    IPv6,
+    Both,
+}
+
+#[derive(Subcommand, Debug, Clone)]
+enum CbfCommands {
+    Rescan {
+        #[clap(long, default_value = "1000")]
+        _batch_size: u32,
+    },
+    Scan,
+    // Used for manual testing as I still haven't implemented Rescan
+    Reveal {
+        #[clap(long, default_value = "1000")]
+        size: u32,
+    },
+}
+
+fn main() -> anyhow::Result<()> {
+    println!("Loading wallet from db...");
+    let (args, keymap, mut keychain_tracker, mut db) =
+        bdk_cli::init::<CbfArgs, CbfCommands, TxHeight>()?;
+    println!("Wallet loaded.");
+
+    let mut client = cbf::CbfClient::new(args.network.into(), args.chain_args.domains)?;
+
+    let cbf_cmd = match args.command {
+        bdk_cli::Commands::ChainSpecific(cbf_cmd) => cbf_cmd,
+        general_cmd => {
+            return bdk_cli::handle_commands(
+                general_cmd,
+                client,
+                &mut keychain_tracker,
+                &mut db,
+                args.network,
+                &keymap,
+            );
+        }
+    };
+
+    match cbf_cmd {
+        CbfCommands::Reveal { size } => {
+            let mut keychain_tracker = keychain_tracker.lock().unwrap();
+            let targets = keychain_tracker
+                .keychains()
+                .into_iter()
+                .map(|(k, _)| (*k, size))
+                .collect();
+            let (_, derivation_additions) = keychain_tracker
+                .txout_index
+                .reveal_to_target_multi(&targets);
+            let changeset = KeychainChangeSet {
+                derivation_indices: derivation_additions,
+                ..Default::default()
+            };
+            db.lock().unwrap().append_changeset(&changeset)?;
+            keychain_tracker.apply_changeset(changeset);
+            Ok(())
+        }
+        CbfCommands::Rescan { _batch_size } => {
+            todo!("Implement rescan from sync");
+            // This function will reveal batch_size scripts and rescan the nakamoto
+            // client
+            // If we notice, after rescan, that more than `batch_size - epsilon` scripts
+            // have been used, we rescan using 2*batch_size, and so on
+        }
+        CbfCommands::Scan => {
+            // indexing logic!
+            let mut keychain_tracker = keychain_tracker.lock().unwrap();
```

Comment:
Try and do every `lock` in scoped block `{...}` so the reader can see where locks can be taken and for how long they need to be taken for. Here you should only need the lock for the time it takes to get the scripts you're interested in and the checkpoints. You can return them from the block and release the lock.